src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,253 @@
+
+    /// <summary>
+    /// Flag that represents whether the LSP view of the world has changed.
+    /// It is totally fine for this to somewhat over-report changes
+    /// as it is an optimization used to delay closing workspace diagnostic requests
+    /// until something has changed.
+    /// </summary>
+    private int _lspChanged = 0;
+
+    protected AbstractWorkspacePullDiagnosticsHandler(
+        LspWorkspaceManager workspaceManager,
+        LspWorkspaceRegistrationService registrationService,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IDiagnosticsRefresher diagnosticRefresher,
+        IGlobalOptionService globalOptions) : base(diagnosticAnalyzerService, diagnosticRefresher, globalOptions)
+    {
+        _workspaceManager = workspaceManager;
+        _workspaceRegistrationService = registrationService;
+
+        _workspaceRegistrationService.LspSolutionChanged += OnLspSolutionChanged;
+        _workspaceManager.LspTextChanged += OnLspTextChanged;
+    }
+
+    public void Dispose()
+    {
+        _workspaceManager.LspTextChanged -= OnLspTextChanged;
+        _workspaceRegistrationService.LspSolutionChanged -= OnLspSolutionChanged;
+    }
+
+    protected override async ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(TDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
+    {
+        // If we're being called from razor, we do not support WorkspaceDiagnostics at all.  For razor, workspace
+        // diagnostics will be handled by razor itself, which will operate by calling into Roslyn and asking for
+        // document-diagnostics instead.
+        if (context.ServerKind == WellKnownLspServerKinds.RazorLspServer)
+            return ImmutableArray<IDiagnosticSource>.Empty;
+
+        var category = GetDiagnosticCategory(diagnosticsParams);
+
+        if (category == PullDiagnosticCategories.Task)
+            return GetTaskListDiagnosticSources(context, GlobalOptions);
+
+        // if this request doesn't have a category at all (legacy behavior, assume they're asking about everything).
+        if (category == null || category == PullDiagnosticCategories.WorkspaceDocumentsAndProject)
+            return await GetDiagnosticSourcesAsync(context, GlobalOptions, cancellationToken).ConfigureAwait(false);
+
+        // if it's a category we don't recognize, return nothing.
+        return ImmutableArray<IDiagnosticSource>.Empty;
+    }
+
+    private void OnLspSolutionChanged(object? sender, WorkspaceChangeEventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void OnLspTextChanged(object? sender, EventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void UpdateLspChanged()
+    {
+        Interlocked.Exchange(ref _lspChanged, 1);
+    }
+
+    protected override async Task WaitForChangesAsync(RequestContext context, CancellationToken cancellationToken)
```

Comment:
Do we have any way to ensure this won't GC root a solution for the duration of the call?

Reply:
Yeah looks like this is already handled here - https://github.com/dotnet/roslyn/blob/4203ffd588ff7a0e05117a5c189649fd3dcc3056/src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs#L202